crates/bevy_ecs/src/query/state.rs

```diff
@@ -245,8 +350,85 @@ where
         if filter.archetype_filter_fetch(location.index) {
             Ok(fetch.archetype_fetch(location.index))
         } else {
-            Err(QueryEntityError::QueryDoesNotMatch)
+            Err(QueryEntityError::QueryDoesNotMatch(entity))
+        }
+    }
+
+    /// Gets the read-only query results for the given [`World`] and array of [`Entity`], where the last change and
+    /// the current change tick are given.
+    pub(crate) fn get_multiple_read_only_manual<'s, 'w, const N: usize>(
+        &'s self,
+        world: &'w World,
+        entities: [Entity; N],
+        last_change_tick: u32,
+        change_tick: u32,
+    ) -> Result<[<Q::ReadOnlyFetch as Fetch<'w, 's>>::Item; N], QueryEntityError> {
+        self.validate_world(world);
+
+        // SAFE: fetch is read-only
+        let array_of_results = unsafe {
+            entities.map(|entity| {
+                self.get_unchecked_manual::<Q::ReadOnlyFetch>(
+                    world,
+                    entity,
+                    last_change_tick,
+                    change_tick,
+                )
+            })
+        };
+
+        // TODO: Replace with TryMap once https://github.com/rust-lang/rust/issues/79711 is stabilized
+        // If any of the get calls failed, bubble up the error
+        for result in &array_of_results {
+            match result {
+                Ok(_) => (),
+                Err(error) => return Err(*error),
+            }
+        }
+
+        // Since we have verified that all entities are present, we can safely unwrap
+        Ok(array_of_results.map(|result| result.unwrap()))
+    }
+
+    /// Gets the query results for the given [`World`] and array of [`Entity`], where the last change and
+    /// the current change tick are given.
+    ///
+    /// # Safety
+    ///
+    /// This does not check for unique access to subsets of the entity-component data.
+    /// To be safe, make sure mutable queries have unique access to the components they query.
+    pub(crate) unsafe fn get_multiple_unchecked_manual<'s, 'w, const N: usize>(
+        &'s self,
+        world: &'w World,
+        entities: [Entity; N],
+        last_change_tick: u32,
+        change_tick: u32,
+    ) -> Result<[<Q::Fetch as Fetch<'w, 's>>::Item; N], QueryEntityError> {
+        self.validate_world(world);
```

Comment:
The `_unchecked_manual` variants don't validate World in any other cases. This saves us on a _lot_ of redundant work, as functions like `Query::get()` might get called hundreds of thousands of times, even in a single system. Instead, we (currently) put the burden on the scheduler to do this.

Sadly I didn't include this in the Safety docs in QueryState (or in `System::run_unsafe`) _and_ theres a gap here because the "safe" `System::run` doesn't validate world.

Obviously thats not good. But I'd prefer to solve the problem holistically / leave the current pattern in-tact. I think validating world at the `System::run_unsafe` level (and updating the `_unchecked_manual` safety docs) has my preference here, given that we are embracing the "just run a system" pattern.

Barring any conflicting thoughts on this (this feels like it will be a divisive topic), can you move `validate_world` "up a level" and out of the `_unchecked_manual` variants?

Reply:
All of the other `_manual` stuff validates world so I think it would be weird for this PR to go and update everything to work differently

Reply:
`manual` validates. `manual_unchecked` does not. I'm only talking about manual_unchecked here.

Reply:
Oops: `unchecked_manual`, not `manual_unchecked`

Reply:
"manual" means you're on the hook to update archetypes (but by default should be assumed to be "safe"). "unchecked" is "fast / unvalidated / unsafe"

Reply:
I'm definitely not asking for changes to unrelated code here. Just for consistency (despite the fact that the pattern being used has some holes, as previously stated).

Reply:
Hmm okay yeah I guess we are missing a lot of methods here so its hard to see a pattern :rofl: `_unchecked` not validating world seems fine I guess :thinking:

Reply:
> Barring any conflicting thoughts on this (this feels like it will be a divisive topic), can you move validate_world "up a level" and out of the _unchecked_manual variants?

Yep, can do.

Reply:
(leaving unresolved so the link in #4363 works nicely)